File: 112. Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// not necessarily non negative numbers, to the leaf 
class Solution {
    //private boolean res = false;
    
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) return false;
        //if (sum == 0) return false;
        return dfs(root,0,sum,root);
    }
    
    public boolean dfs(TreeNode root, int temp, int sum, TreeNode parent) {
        if (temp == sum && root == null && parent.left == null && parent.right == null) {
            return true;
        }
        if (root == null) return false; 
        //boolean isLeaf = root.left == null && root.right == null;
        
        boolean hasLeft = dfs(root.left, temp + root.val,sum, root);
        boolean hasRight = dfs(root.right,temp + root.val, sum, root);
        return hasLeft || hasRight;
    }
}

File: amazon/python/112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        if not root:
            return False
        sum -= root.val # important to remember 
        if (not root.left) and (not root.right):
            return sum == 0
        return self.hasPathSum(root.left, sum) or self.hasPathSum(root.right, sum)
        
    # Time: O(n)
    # Space: O(n)
    
#     def hasPathSum(self, root, sum):
#         """
#         :type root: TreeNode
#         :type sum: int
#         :rtype: bool
#         """
#         # return all root to leaf combinations 
#         # store all node and parent information 
#         if not root:
#             return False
#         stack = [(root,sum-root.val)]
        
#         while stack:
#             cur, remain = stack.pop()
#             if remain == 0 and (not cur.left) and (not cur.right):
#                 return True
#             else:
#                 if cur.left:
#                     stack.append((cur.left, remain - cur.left.val))
#                 if cur.right:
#                     stack.append((cur.right,remain - cur.right.val))
#         return False 
#     # Time: O(n)
#     # Space: O(n)
    
            
